// react-mutation-observer/lib/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', './withObserver', './WatchForRemoval', './WatchForChildrenAddition', './WatchChildren', './WatchAttributes', './WatchContent', './WatchForChildrenRemoval', './MutationObserver'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('./withObserver'), require('./WatchForRemoval'), require('./WatchForChildrenAddition'), require('./WatchChildren'), require('./WatchAttributes'), require('./WatchContent'), require('./WatchForChildrenRemoval'), require('./MutationObserver'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.withObserver, global.WatchForRemoval, global.WatchForChildrenAddition, global.WatchChildren, global.WatchAttributes, global.WatchContent, global.WatchForChildrenRemoval, global.MutationObserver);
    global.index = mod.exports;
  }
})(this, function (exports, _withObserver, _WatchForRemoval, _WatchForChildrenAddition, _WatchChildren, _WatchAttributes, _WatchContent, _WatchForChildrenRemoval, _MutationObserver) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.CHARACTER_DATA = exports.CHILD_REMOVED = exports.CHILD_ADDED = exports.ATTRIBUTES = exports.CHILD_LIST = exports.WatchContent = exports.WatchAttributes = exports.WatchChildren = exports.WatchForChildrenRemoval = exports.WatchForChildrenAddition = exports.WatchForRemoval = exports.MutationObserver = exports.withObserver = undefined;

  var _withObserver2 = _interopRequireDefault(_withObserver);

  var _WatchForRemoval2 = _interopRequireDefault(_WatchForRemoval);

  var _WatchForChildrenAddition2 = _interopRequireDefault(_WatchForChildrenAddition);

  var _WatchChildren2 = _interopRequireDefault(_WatchChildren);

  var _WatchAttributes2 = _interopRequireDefault(_WatchAttributes);

  var _WatchContent2 = _interopRequireDefault(_WatchContent);

  var _WatchForChildrenRemoval2 = _interopRequireDefault(_WatchForChildrenRemoval);

  var _MutationObserver2 = _interopRequireDefault(_MutationObserver);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports.default = _MutationObserver2.default;
  exports.withObserver = _withObserver2.default;
  exports.MutationObserver = _MutationObserver2.default;
  exports.WatchForRemoval = _WatchForRemoval2.default;
  exports.WatchForChildrenAddition = _WatchForChildrenAddition2.default;
  exports.WatchForChildrenRemoval = _WatchForChildrenRemoval2.default;
  exports.WatchChildren = _WatchChildren2.default;
  exports.WatchAttributes = _WatchAttributes2.default;
  exports.WatchContent = _WatchContent2.default;
  exports.CHILD_LIST = _withObserver.CHILD_LIST;
  exports.ATTRIBUTES = _withObserver.ATTRIBUTES;
  exports.CHILD_ADDED = _withObserver.CHILD_ADDED;
  exports.CHILD_REMOVED = _withObserver.CHILD_REMOVED;
  exports.CHARACTER_DATA = _withObserver.CHARACTER_DATA;
});
